File: app/src/main/java/com/xxx/ware_house/net/RetrofitManager.java
package com.xxx.ware_house.net;

import android.os.Environment;
import android.text.TextUtils;

import com.xxx.ware_house.api.RetrofitApiService;
import com.xxx.ware_house.net.interceptor.HttpLogInterceptor;
import com.xxx.ware_house.net.interceptor.NetCacheInterceptor;
import com.xxx.ware_house.net.interceptor.OfflineCacheInterceptor;
import com.xxx.ware_house.utils.LogUtils;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.Cache;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * @Author: ZhangRuixiang
 * Date: 2021/11/13
 * DES:
 */
public class RetrofitManager {

    private static RetrofitManager retrofitManager;
    private OkHttpClient okHttpClient;
    private Retrofit retrofit;
    private ArrayList<String> oneNetList;
    private static String baseUrl;
    private static List<String> imageUrlList;
    private static Class<?> retrofitApiServiceClass;
    private RetrofitApiService retrofitApiService;

    private RetrofitManager() {
        oneNetList = new ArrayList<>();
        initOkHttpClient();
        initRetrofit();
    }

    /**
     * 第一次或者重置baseUrl时使用
     * 建议在application中初始化
     * @param baseUrl 不可空
     * @param list  不可空
     */
    public static void getInstance(String baseUrl, List<String> list, Class<?> retrofitApiService) {
        if (retrofitManager == null) {
            synchronized (RetrofitManager.class) {
                if (retrofitManager == null) {
                    RetrofitManager.baseUrl = baseUrl;
                    RetrofitManager.imageUrlList = list;
                    RetrofitManager.retrofitApiServiceClass = retrofitApiService;
                    retrofitManager = new RetrofitManager();
                } else {
                    retrofitManager = null;
                    getInstance(baseUrl, list,retrofitApiService);
                }
            }
        }else {
            retrofitManager = null;
            getInstance(baseUrl, list,retrofitApiService);
        }
    }

    public static RetrofitManager getInstance(){
        if (TextUtils.isEmpty(baseUrl)){
            LogUtils.e(RetrofitManager.class.getSimpleName(),"baseUrl为空，无法初始化");
            return null;
        }
        if (retrofitManager == null){
            synchronized (RetrofitManager.class){
                if (retrofitManager == null){
                    retrofitManager = new RetrofitManager();
                }
            }
        }
        return retrofitManager;
    }

    private void initRetrofit() {
        retrofit = new Retrofit.Builder()
                .baseUrl(baseUrl)
                .addConverterFactory(GsonConverterFactory.create())
                .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
                .client(okHttpClient)
                .build();
        retrofitApiService = (RetrofitApiService) retrofit.create(retrofitApiServiceClass);
    }

    private void initOkHttpClient() {
        okHttpClient = new OkHttpClient.Builder()
                .cache(new Cache(new File(Environment.getExternalStorageDirectory() + "/okhttp_cache/"), 50 * 1024 * 1024))
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(10, TimeUnit.SECONDS)
                .writeTimeout(10, TimeUnit.SECONDS)
                .addInterceptor(new HttpLogInterceptor(imageUrlList))
                //设置在线和离线缓存
                .addInterceptor(OfflineCacheInterceptor.getInstance())
                .addNetworkInterceptor(NetCacheInterceptor.getInstance())
                .build();
    }

    public  ArrayList<String> getOneNetList(){
        return oneNetList;
    }

    public RetrofitApiService getRetrofitApiService(){
        if (retrofitManager == null){
            retrofitManager = getInstance();
        }
        assert retrofitManager != null;
        return retrofitManager.retrofitApiService;
    }

}

File: app/src/main/java/com/xxx/ware_house/net/RetrofitManager.kt
package com.xxx.ware_house.net

import android.os.Environment
import android.text.TextUtils
import com.xxx.ware_house.api.RetrofitApiService
import com.xxx.ware_house.common.SystemCommon
import com.xxx.ware_house.net.interceptor.HttpLogInterceptor
import com.xxx.ware_house.net.interceptor.NetCacheInterceptor
import com.xxx.ware_house.net.interceptor.OfflineCacheInterceptor
import com.xxx.ware_house.utils.LogUtils
import com.xxx.ware_house.utils.SpUtil
import okhttp3.Cache
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File
import java.util.concurrent.TimeUnit

/**
 * @Author: ZhangRuixiang
 * Date: 2021/11/13
 * DES:
 */
class RetrofitManager private constructor() {
    private var okHttpClient: OkHttpClient? = null
    private var retrofit: Retrofit? = null
    val oneNetList: ArrayList<String>
    private var retrofitApiService: RetrofitApiService? = null

    init {
        oneNetList = ArrayList()
        initOkHttpClient()
        initRetrofit()
    }

    private fun initRetrofit() {
        retrofit = Retrofit.Builder()
            .baseUrl(baseUrl)
            .addConverterFactory(GsonConverterFactory.create())
            .addCallAdapterFactory(RxJava2CallAdapterFactory.create())
            .client(okHttpClient)
            .build()
        retrofitApiService = retrofit?.create(retrofitApiServiceClass) as RetrofitApiService
    }

    private fun initOkHttpClient() {
        okHttpClient = OkHttpClient.Builder()
            .cache(
                Cache(
                    File(
                        Environment.getExternalStorageDirectory().toString() + "/okhttp_cache/"
                    ), (50 * 1024 * 1024).toLong()
                )
            )
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(10, TimeUnit.SECONDS)
            .writeTimeout(10, TimeUnit.SECONDS)
            .addInterceptor(HttpLogInterceptor(imageUrlList)) //设置在线和离线缓存
            .addInterceptor(OfflineCacheInterceptor.getInstance())
            .addNetworkInterceptor(NetCacheInterceptor.getInstance())
            .build()
    }

    fun getRetrofitApiService(): RetrofitApiService? {
        if (retrofitManager == null) {
            retrofitManager = instance
        }
        assert(retrofitManager != null)
        return retrofitManager!!.retrofitApiService
    }

    companion object {
        private var retrofitManager: RetrofitManager? = null
        private var baseUrl: String? = null
        private var imageUrlList: List<String>? = null
        private var retrofitApiServiceClass: Class<*>? = null

        /**
         * 第一次或者重置baseUrl时使用
         * 建议在application中初始化
         * @param baseUrl 不可空
         * @param list  不可空
         */
        fun getInstance(baseUrl: String?, list: List<String>?, retrofitApiService: Class<*>?) {
            if (retrofitManager == null) {
                synchronized(RetrofitManager::class.java) {
                    if (retrofitManager == null) {
                        Companion.baseUrl = baseUrl
                        imageUrlList = list
                        retrofitApiServiceClass = retrofitApiService
                        retrofitManager = RetrofitManager()
                    } else {
                        retrofitManager = null
                        getInstance(baseUrl, list, retrofitApiService)
                    }
                }
            } else {
                retrofitManager = null
                getInstance(baseUrl, list, retrofitApiService)
            }
        }

        val instance: RetrofitManager?
            get() {
                if (TextUtils.isEmpty(baseUrl)) {
                    LogUtils.e(RetrofitManager::class.java.simpleName, "baseUrl为空，无法初始化")
                    return null
                }
                if (retrofitManager == null) {
                    synchronized(RetrofitManager::class.java) {
                        if (retrofitManager == null) {
                            retrofitManager = RetrofitManager()
                        }
                    }
                }
                return retrofitManager
            }
    }
}